// next/dist/client/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderError = exports.render = exports.router = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var render = exports.render = function () {
  var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(props) {
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            if (!props.err) {
              _context.next = 4;
              break;
            }

            _context.next = 3;
            return renderError(props.err);

          case 3:
            return _context.abrupt('return');

          case 4:
            _context.prev = 4;
            _context.next = 7;
            return doRender(props);

          case 7:
            _context.next = 15;
            break;

          case 9:
            _context.prev = 9;
            _context.t0 = _context['catch'](4);

            if (!_context.t0.abort) {
              _context.next = 13;
              break;
            }

            return _context.abrupt('return');

          case 13:
            _context.next = 15;
            return renderError(_context.t0);

          case 15:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[4, 9]]);
  }));

  return function render(_x) {
    return _ref2.apply(this, arguments);
  };
}();

// This method handles all runtime and debug errors.
// 404 and 500 errors are special kind of errors
// and they are still handle via the main render method.


var renderError = exports.renderError = function () {
  var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(error) {
    var prod, errorMessage, initProps, _props;

    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            prod = process.env.NODE_ENV === 'production';
            // We need to unmount the current app component because it's
            // in the inconsistant state.
            // Otherwise, we need to face issues when the issue is fixed and
            // it's get notified via HMR

            _reactDom2.default.unmountComponentAtNode(appContainer);

            errorMessage = error.message + '\n' + error.stack;

            console.error(errorMessage);

            if (!prod) {
              _context2.next = 12;
              break;
            }

            initProps = { err: error, pathname: pathname, query: query };
            _context2.next = 8;
            return (0, _utils.loadGetInitialProps)(ErrorComponent, initProps);

          case 8:
            _props = _context2.sent;

            _reactDom2.default.render((0, _react.createElement)(ErrorComponent, _props), errorContainer);
            _context2.next = 13;
            break;

          case 12:
            _reactDom2.default.render((0, _react.createElement)(_errorDebug2.default, { error: error }), errorContainer);

          case 13:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function renderError(_x2) {
    return _ref3.apply(this, arguments);
  };
}();

var doRender = function () {
  var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(_ref5) {
    var Component = _ref5.Component,
        props = _ref5.props,
        hash = _ref5.hash,
        err = _ref5.err,
        emitter = _ref5.emitter;

    var _pathname, _query, appProps;

    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            if (!(!props && Component && Component !== ErrorComponent && lastAppProps.Component === ErrorComponent)) {
              _context3.next = 5;
              break;
            }

            // fetch props if ErrorComponent was replaced with a page component by HMR
            _pathname = router.pathname, _query = router.query;
            _context3.next = 4;
            return (0, _utils.loadGetInitialProps)(Component, { err: err, pathname: _pathname, query: _query });

          case 4:
            props = _context3.sent;

          case 5:

            if (emitter) {
              emitter.emit('before-reactdom-render', { Component: Component });
            }

            Component = Component || lastAppProps.Component;
            props = props || lastAppProps.props;

            appProps = { Component: Component, props: props, hash: hash, err: err, router: router, headManager: headManager };
            // lastAppProps has to be set before ReactDom.render to account for ReactDom throwing an error.

            lastAppProps = appProps;

            // We need to clear any existing runtime error messages
            _reactDom2.default.unmountComponentAtNode(errorContainer);
            _reactDom2.default.render((0, _react.createElement)(_app2.default, appProps), appContainer);

            if (emitter) {
              emitter.emit('after-reactdom-render', { Component: Component });
            }

          case 13:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this);
  }));

  return function doRender(_x3) {
    return _ref4.apply(this, arguments);
  };
}();

var _react = require('react');

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _mitt = require('mitt');

var _mitt2 = _interopRequireDefault(_mitt);

var _headManager = require('./head-manager');

var _headManager2 = _interopRequireDefault(_headManager);

var _router = require('../lib/router');

var _app = require('../lib/app');

var _app2 = _interopRequireDefault(_app);

var _evalScript = require('../lib/eval-script');

var _evalScript2 = _interopRequireDefault(_evalScript);

var _utils = require('../lib/utils');

var _errorDebug = require('../lib/error-debug');

var _errorDebug2 = _interopRequireDefault(_errorDebug);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Polyfill Promise globally
// This is needed because Webpack2's dynamic loading(common chunks) code
// depends on Promise.
// So, we need to polyfill it.
// See: https://github.com/webpack/webpack/issues/4254
if (!window.Promise) {
  window.Promise = _promise2.default;
}

var _window = window,
    _window$__NEXT_DATA__ = _window.__NEXT_DATA__,
    component = _window$__NEXT_DATA__.component,
    errorComponent = _window$__NEXT_DATA__.errorComponent,
    props = _window$__NEXT_DATA__.props,
    err = _window$__NEXT_DATA__.err,
    pathname = _window$__NEXT_DATA__.pathname,
    query = _window$__NEXT_DATA__.query,
    location = _window.location;


var Component = (0, _evalScript2.default)(component).default;
var ErrorComponent = (0, _evalScript2.default)(errorComponent).default;
var lastAppProps = void 0;

var router = exports.router = (0, _router.createRouter)(pathname, query, (0, _utils.getURL)(), {
  Component: Component,
  ErrorComponent: ErrorComponent,
  err: err
});

var headManager = new _headManager2.default();
var appContainer = document.getElementById('__next');
var errorContainer = document.getElementById('__next-error');

exports.default = function () {
  var emitter = (0, _mitt2.default)();

  router.subscribe(function (_ref) {
    var Component = _ref.Component,
        props = _ref.props,
        hash = _ref.hash,
        err = _ref.err;

    render({ Component: Component, props: props, err: err, hash: hash, emitter: emitter });
  });

  var hash = location.hash.substring(1);
  render({ Component: Component, props: props, hash: hash, err: err, emitter: emitter });

  return emitter;
};
